### PR TITLE
Nano: add channels_last  convert check

### PR DESCRIPTION
**Pull request by ACupofAir**
### 1. Why the change?
When we pass in a `input_sample` which is not consist of tensor, or one of the item in input doesn't have `.to()` method, it will lead to an error to convert it on `channels_last` is true.

### 2. Summary of the change 
* When user directly use `trace` or `quantize` function with a `input_sample` pass in or use `optimize` function which mean the `input_sample` is **not None**, we will compute a list named `channels_last_available` which instruct which item in the `input_sample` can be convert to `channels_last`.
* When user directly use `trace` or `quantize` function and don't pass in `input_sample`, I make the list `channels_last_available` empty, and the `forward_step` function will convert default, but if there are some errors occure, there is an alert to tell user they should pass in a `input_sample` to try to solve it.

### CODE REVIEW

File: python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py [hunk 1]
```diff
@@ -94,16 +110,28 @@ def __init__(self, model: torch.nn.Module, input_sample=None, use_ipex=False, dt
...
+        ]
 
     def on_forward_start(self, inputs):
         return inputs
 
     def forward_step(self, *inputs):
         if self.channels_last is True:
-            inputs = tuple(map(lambda x: x.to(memory_format=torch.channels_last), inputs))
-        return self.model(*inputs)
+            if self.channels_last_available: # it mean the input_sample is not None
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py:124]:
```python
# check if `channels_last_available` has been generated
```

File: python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py [hunk 2]
```diff
@@ -94,16 +110,28 @@ def __init__(self, model: torch.nn.Module, input_sample=None, use_ipex=False, dt
...
+                for idx, input in enumerate(inputs):
+                    if self.channels_last_available[idx]:
+                        input.to(memory_format=torch.channels_last)
+            else:
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py:128]: why don't we just generate the `channels_last_available` here? we have a precious input sample right :)

File: python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py [hunk 3]
```diff
@@ -94,16 +110,28 @@ def __init__(self, model: torch.nn.Module, input_sample=None, use_ipex=False, dt
...
+                try:
+                    inputs = tuple(map(lambda x: x.to(memory_format=torch.channels_last), inputs))
+                except:
+                    print("************************error:***************************\n")
+                    print("You should pass in a non_empty input_sample")
+        return self.model(*tuple(inputs))
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py:134]: could we just use `*inputs` here?

File: python/nano/src/bigdl/nano/pytorch/amp/bfloat16.py [hunk 4]
```diff
@@ -106,7 +107,18 @@ def on_forward_start(self, inputs):
...
+            if self.channels_last_available: # it mean the input_sample is not None
+                for idx, input in enumerate(inputs):
+                    if self.channels_last_available[idx]:
+                        input.to(memory_format=torch.channels_last)
+            else:
+                try:
+                    inputs = tuple(map(lambda x: x.to(memory_format=torch.channels_last), inputs))
+                except:
+                    print("************************error:***************************\n")
+                    print("You should pass in a non_empty input_sample")
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/amp/bfloat16.py:119]: same comments as before

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 5]
```diff
@@ -560,6 +560,19 @@ def quantize(model: nn.Module,
...
+        channels_last_available = []
+        if input_sample:
+            channels_last_available = [True]*len(input_sample)
+            for idx in range(len(input_sample)):
+                try:
+                    input_sample[idx].to(memory_format=torch.channels_last)
+                except:
+                    channels_last_available[idx] = False
+                else:
+                    channels_last_available[idx] = True
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:574]: we should move this logic to `PytorchIPEXJITModel`'s `__init__()`

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 6]
```diff
@@ -753,6 +769,20 @@ def trace(model: nn.Module,
...
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:783]: same here

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 7]
```diff
@@ -780,6 +810,7 @@ def trace(model: nn.Module,
             use_jit = (accelerator == "jit")
             return PytorchIPEXJITModel(model, input_sample=input_sample, use_ipex=use_ipex,
                                        use_jit=use_jit, channels_last=channels_last,
+                                       channels_last_available=channels_last_available,
                                        thread_num=thread_num, inplace=inplace,
                                        jit_strict=jit_strict)
         invalidInputError(False, "Accelerator {} is invalid.".format(accelerator))
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:816]: we want some unit test for this fix in https://github.com/intel-analytics/BigDL/blob/main/python/nano/test/pytorch/tests/inference/test_ipex_jit_inference.py

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:816]: we need to make sure `channels_last_available` is well recovered even after `InferenceOptimizer.save/load`

File: python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py [hunk 8]
```diff
@@ -117,18 +157,22 @@ def __getattr__(self, name: str):
...
-                       "thread_num": self.thread_num,
-                       "jit_strict": self.jit_strict})
+        status.update(
+            {
+                "use_ipex": self.use_ipex,
+                "use_jit": self.use_jit,
+                "channels_last": self.channels_last,
+                "checkpoint": "ckpt.pth",
+                "thread_num": self.thread_num,
+                "jit_strict": self.jit_strict,
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py:167]: should we save the `self.channels_last_available` here? Since we do want it be recoverred after users calling save/load

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 9]
```diff
@@ -350,6 +350,8 @@ def optimize(self, model: nn.Module,
                                                                thread_num=thread_num)
 
         print("==========================Start Optimization==========================")
+        print('======================DEBUG START: Junes nano======================')
+        print('======================DEBUG  END : Junes nano======================')
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:354]: no need for these 2 lines

File: python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py (hunk 2 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/deps/ipex/ipex_inference_model.py:135]: change to something like
```python
else:
    self.channels_last_available = generate_channels_last_available(inputs)
    for idx, input in enumerate(inputs):
        if self.channels_last_available[idx]:
            input.to(memory_format=torch.channels_last)
```

File: python/nano/test/pytorch/tests/inference/test_bf16_ipex.py [hunk 10]
```diff
@@ -151,6 +152,116 @@ def hello():
             new_model.width
 
 
+    def test_bf16_channels_last_inference_with_jit(self):
+        model = resnet18(num_classes=10)
+
+        x = torch.rand((10, 3, 256, 256))
```
Comment (TheaperDeng) [inline python/nano/test/pytorch/tests/inference/test_bf16_ipex.py:158]: it seems we keep using `x=torch.rand((10, 3, 256, 256))` as the test input sample. An easy way to do this could be
1. add 1 test function in each test file (test_bf16, test_bf16_ipex, test_ipex_jit_inference)
2. For these test function, write your own test model (could be an extremely easy one with 3 input)
3. use (x1, x2, x3) as input, where x1, x3 are 4-dim tensor while x2 is 3-dim.
4. test save/load

**Comment by ACupofAir** [comment]
replaced by #7151
